File: reference/src/core/validation.ts
/**
 * Schema validation using ajv against AI-SDLC JSON Schema definitions.
 * Uses ajv/dist/2020 for JSON Schema draft 2020-12 support.
 */

import _Ajv2020 from 'ajv/dist/2020.js';
import _addFormats from 'ajv-formats';
import { commonSchema, SCHEMAS } from './generated-schemas.js';
import type { ResourceKind, AnyResource } from './types.js';

// Handle CJS default export interop
const Ajv2020 = _Ajv2020 as unknown as typeof _Ajv2020.default;
const addFormats = _addFormats as unknown as typeof _addFormats.default;

export interface ValidationResult<T = AnyResource> {
  valid: boolean;
  data?: T;
  errors?: ValidationError[];
}

export interface ValidationError {
  path: string;
  message: string;
  keyword: string;
}

const SCHEMA_FILES: Record<ResourceKind, string> = {
  Pipeline: 'pipeline.schema.json',
  AgentRole: 'agent-role.schema.json',
  QualityGate: 'quality-gate.schema.json',
  AutonomyPolicy: 'autonomy-policy.schema.json',
  AdapterBinding: 'adapter-binding.schema.json',
  DesignSystemBinding: 'design-system-binding.schema.json',
  DesignIntentDocument: 'design-intent-document.schema.json',
};

type AjvInstance = InstanceType<typeof Ajv2020>;
type ValidatorFn = ReturnType<AjvInstance['compile']>;

let ajvInstance: AjvInstance | null = null;
const validators = new Map<ResourceKind, ValidatorFn>();

function getAjv(): AjvInstance {
  if (!ajvInstance) {
    ajvInstance = new Ajv2020({
      allErrors: true,
      strict: false,
    });
    addFormats(ajvInstance);

    // Load common schema from inline generated module
    ajvInstance.addSchema(commonSchema);
  }
  return ajvInstance;
}

function getValidator(kind: ResourceKind): ValidatorFn {
  let validator = validators.get(kind);
  if (!validator) {
    const schemaFile = SCHEMA_FILES[kind];
    if (!schemaFile) {
      throw new Error(`Unknown resource kind: ${kind}`);
    }
    const schema = SCHEMAS[schemaFile];
    if (!schema) {
      throw new Error(`Schema not found for: ${schemaFile}`);
    }
    validator = getAjv().compile(schema);
    validators.set(kind, validator);
  }
  return validator;
}

/**
 * Collapse noisy AJV errors — especially oneOf branch failures — into
 * a concise list.  AJV reports individual branch errors (schemaPath
 * contains `/oneOf/0/`, `/oneOf/1/`, …) AND a summary `keyword:
 * 'oneOf'` error.  We keep the summary and drop the per-branch noise.
 */
export function formatValidationErrors(
  rawErrors: Array<{
    instancePath: string;
    message?: string;
    keyword: string;
    params?: Record<string, unknown>;
    schemaPath: string;
  }>,
): ValidationError[] {
  const errors: ValidationError[] = [];

  for (const err of rawErrors) {
    // Skip individual oneOf branch failures — they have schemaPath containing "/oneOf/"
    if (err.schemaPath.includes('/oneOf/') && err.keyword !== 'oneOf') {
      continue;
    }

    // For the summary "oneOf" error itself, provide a clearer message
    if (err.keyword === 'oneOf') {
      errors.push({
        path: err.instancePath || '/',
        message:
          'Value must match exactly one of the allowed variants (check the schema for valid shapes)',
        keyword: 'oneOf',
      });
      continue;
    }

    // Similarly collapse anyOf branch noise
    if (err.schemaPath.includes('/anyOf/') && err.keyword !== 'anyOf') {
      continue;
    }
    if (err.keyword === 'anyOf') {
      errors.push({
        path: err.instancePath || '/',
        message: 'Value must match at least one of the allowed variants',
        keyword: 'anyOf',
      });
      continue;
    }

    errors.push({
      path: err.instancePath || '/',
      message: err.message ?? 'Unknown validation error',
      keyword: err.keyword,
    });
  }

  return errors;
}

/**
 * Validate a resource document against its JSON Schema.
 */
export function validate<T extends AnyResource = AnyResource>(
  kind: ResourceKind,
  data: unknown,
): ValidationResult<T> {
  const validator = getValidator(kind);
  const valid = validator(data);

  if (valid) {
    return { valid: true, data: data as T };
  }

  const errors = formatValidationErrors(validator.errors ?? []);

  return { valid: false, errors };
}

/**
 * Validate a resource, inferring the kind from the document's `kind` field.
 */
export function validateResource(data: unknown): ValidationResult {
  if (typeof data !== 'object' || data === null || !('kind' in data)) {
    return {
      valid: false,
      errors: [{ path: '/', message: 'Missing "kind" field', keyword: 'required' }],
    };
  }

  const kind = (data as { kind: string }).kind as ResourceKind;
  if (!(kind in SCHEMA_FILES)) {
    return {
      valid: false,
      errors: [{ path: '/kind', message: `Unknown resource kind: ${kind}`, keyword: 'enum' }],
    };
  }

  return validate(kind, data);
}

File: sdk-python/src/ai_sdlc/core/validation.py
"""Schema validation using jsonschema against AI-SDLC JSON Schema definitions."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import TYPE_CHECKING, Any

import jsonschema
import jsonschema.validators
from referencing import Registry, Resource

if TYPE_CHECKING:
    from .types import ResourceKind

_SCHEMA_DIR = Path(__file__).resolve().parent.parent.parent.parent.parent / "spec" / "schemas"

SCHEMA_FILES: dict[str, str] = {
    "Pipeline": "pipeline.schema.json",
    "AgentRole": "agent-role.schema.json",
    "QualityGate": "quality-gate.schema.json",
    "AutonomyPolicy": "autonomy-policy.schema.json",
    "AdapterBinding": "adapter-binding.schema.json",
}


@dataclass(frozen=True)
class ValidationError:
    path: str
    message: str
    keyword: str


@dataclass
class ValidationResult:
    valid: bool
    data: dict[str, Any] | None = None
    errors: list[ValidationError] = field(default_factory=list)


def _load_schema(filename: str) -> dict[str, Any]:
    import json

    return json.loads((_SCHEMA_DIR / filename).read_text())  # type: ignore[no-any-return]


def _build_registry() -> Registry[dict[str, Any]]:
    """Build a referencing.Registry with all schemas pre-loaded."""
    import json

    resources: list[tuple[str, Resource[dict[str, Any]]]] = []
    for p in _SCHEMA_DIR.glob("*.schema.json"):
        schema = json.loads(p.read_text())
        schema_id = schema.get("$id", "")
        resources.append((schema_id, Resource.from_contents(schema)))
    return Registry[dict[str, Any]]().with_resources(resources)


_registry: Registry[dict[str, Any]] | None = None


def _get_registry() -> Registry[dict[str, Any]]:
    global _registry  # noqa: PLW0603
    if _registry is None:
        _registry = _build_registry()
    return _registry


def _get_validator(kind: str) -> jsonschema.Validator:
    schema_file = SCHEMA_FILES.get(kind)
    if not schema_file:
        raise ValueError(f"Unknown resource kind: {kind}")
    schema = _load_schema(schema_file)
    validator_cls = jsonschema.validators.validator_for(schema)
    return validator_cls(schema, registry=_get_registry())


def validate(kind: ResourceKind, data: Any) -> ValidationResult:
    """Validate a resource document against its JSON Schema."""
    validator = _get_validator(kind)
    errors: list[ValidationError] = []
    for error in validator.iter_errors(data):
        path = "/" + "/".join(str(p) for p in error.absolute_path) if error.absolute_path else "/"
        errors.append(ValidationError(
            path=path,
            message=error.message,
            keyword=str(error.validator),
        ))

    if errors:
        return ValidationResult(valid=False, errors=errors)
    return ValidationResult(valid=True, data=data)


def validate_resource(data: Any) -> ValidationResult:
    """Validate a resource, inferring the kind from the document's ``kind`` field."""
    if not isinstance(data, dict) or "kind" not in data:
        return ValidationResult(
            valid=False,
            errors=[ValidationError(path="/", message='Missing "kind" field', keyword="required")],
        )

    kind = data["kind"]
    if kind not in SCHEMA_FILES:
        return ValidationResult(
            valid=False,
            errors=[
                ValidationError(
                    path="/kind",
                    message=f"Unknown resource kind: {kind}",
                    keyword="enum",
                )
            ],
        )
    return validate(kind, data)
